### ISSUE TITLE
Cannot reorder admin tabs

### ISSUE THREAD

**Issue by mlinfoot** [refinery/refinerycms#3024]
Upgrading to latest commit (9a69a74) resulted in no way to reorder admin tabs (Users, Pages, etc.). The reorder icon visible in previous versions no longer appears.

Reproduced with a clean, fresh install --  rails new myapp -m http://refinerycms.com/t/edge

**Comment by anitagraham** [comment]
Confirmed.
Now investigating

**Comment by anitagraham** [comment]
First observation:
the code for adding the reorder/reorder_done items was removed from `core/app/views/refinery/admin/_menu.html.erb` in [Removed sortable menu feature, this belongs in an authentication plugin.](https://github.com/refinery/refinerycms/commit/2c2983aa2aab551aaaedcade3511830b8193f023).

Now to see where they have gone.

**Comment by parndt** [comment]
I can't be sure that I put it in

**Comment by anitagraham** [comment]
After further thought, I wonder why the sortable menu feature belongs in authentication.
Perhaps being able to sort menus is something that should require authentication - but doesn't the actual sorting belong just where it was?
What role would a user need to be able to sort the admin tabs?

**Comment by parndt** [comment]
@anitagraham the issue is that there's no concept of a user at all in the core system anymore so to persist a personalised order dosn't make sense without an authentication system with users.

**Comment by bricesanchez** [comment]
:+1 @parndt

**Comment by anitagraham** [comment]
Does the order need to be personalised? 

I would suggest that reordering could be done (and persisted at a global level) by anyone with appropriate authorization? 

How would it need to work within an authentication plugin?

Do you mean in a newly conceived authentication plugin or just within one of the already existing ones.

I guess my question might translate to "What is the authentication API?".
(We've been talking API design at home. SO met someone who teaches API design last week.)

**Comment by parndt** [comment]
I guess I just think it's outside the scope of the core system

**Comment by anitagraham** [comment]
Not sure I agree, tbh.

**Comment by parndt** [comment]
the authentication/authorisation system then we can fully outsource the burden of maintenance as the way it works will be different for each extension.

**Comment by anitagraham** [comment]
I don't know enough about refinery or the other possible authentication systems to point a finger and say "there". Previously this was a function that (afaik) was available to any user.  Can't we leave it like that for the moment (as said in gitter by @mlinfoot ).

Where do I read about refinery authentication - users, roles, privileges/actions?

**Comment by prokopsimek** [comment]
I have two projects...
1st is bundled to commit "29ab03669a0bdba34c789442df1402f6df06e929" - reordering fails
2nd is bundled to commit "0bbbc54f7910b8a5436ed7351bb4efb35f6f4c20" - reorderign is ok

but if I updated 1st app reference to commit the same as the second, reordering still fails... No JS error, no console error... Nothing...

**Comment by bricesanchez** [comment]
The problem with tabs reordering is if it's not in core, we will have to copy/paste the feature in each authentication system (like for spree-refinerycms-authentication). Maybe we should leave it in the core or create an engine for this feature :(

**Comment by mlinfoot** [comment]
@bricesanchez +1 for leaving it in core.

**Comment by gwagener** [comment]
My few cents. 1) I always expected it to be site wide, not user specific, so I would be fine with it being thus limited and in core. 2) It's functionality for organising plugins so it probably belongs where the support for a pluggable architecture is which is presumably core. 3) You could always add screeds of complexity and have basic, non-user specific functionality in core that is then overridden, or not, with user specific functionality in authentication extensions. Lol complexity! 4) It's really not a necessary feature. 5) Ignore me because I don't have any skin in the game.

**Comment by parndt** [comment]
Thanks @gwagener I largely agree with point 4

I like this solution best:

```ruby
Refinery::Core.config.plugin_order = %w(refinery_pages refinery_images)
```

Anything not in that list just gets chucked at the end like it does now but anything _in_ the list gets prioritised first.

**Comment by parndt** [comment]
So we want something like this (suggested by @gwagener):

```ruby
Refinery::Core.plugin_order = %w(refinery_pages refinery_images)
```

And in the [plugin code in_menu method](https://github.com/refinery/refinerycms/blob/f9c15e8a4478b73a4115095cec545912ad6dd8f6/core/lib/refinery/plugins.rb#L23-L26) can change from this:

```ruby
# TODO: Review necessary?
def in_menu
  self.class.new(reject(&:hide_from_menu))
end
```

To this:

```ruby
def in_menu
  sorted = Refinery::Core.config.plugin_order
  items_in_menu = reject(&:hide_from_menu)
  self.class.new(sorted + (items_in_menu - sorted))
end
```

This is pseudocode and will need to be turned into something that performs this operation on the plugin's name.. but that is how we'd support it.

Once we get this in, we can ship 3.0 right?

**Comment by gwagener** [comment]
I realised it has to be `(sorted & items_in_menu) + (items_in_menu - sorted)`.

**Comment by bricesanchez** [comment]
I'm agree with @gwagener and @parndt for this feature :)

**Comment by parndt** [comment]
This is fixed with #3053
